tensorflow/contrib/periodic_intersperse/core/kernels/periodic_intersperse_op.h

```diff
@@ -0,0 +1,231 @@
+  // factor by which X increases/decreases w.r.t. Y
+  std::vector<float> f(rank);
+  // helper arrays related to f
+  std::vector<tensorflow::int64> g(rank), G(rank);
+  // index of adjustable dimension
+  int q;
+  tensorflow::TensorShape output_shape;
+
+  // requires that the rank of the input tensor and length of the desired shape
+  // are equal
+  OP_REQUIRES(context, rank == desired_shape.size(),
+              tensorflow::errors::InvalidArgument(
+                  "periodic_intersperse expects the rank of the input tensor, ",
+                  rank, ", to be the same as the length of the desired shape, ",
+                  desired_shape.size(), "."));
+
+  {
+    bool found = false;
+    for (int i = 0; i < rank; ++i) {
+      if (desired_shape(i) < 1) {
+        // only one index can be adjustable
+        OP_REQUIRES(context, !found,
+                    tensorflow::errors::InvalidArgument(
+                        "periodic_intersperse expects only "
+                        "one index to be marked as adjustable."));
+        q = i;
+        found = true;
+      } else {
+        Y[i] = desired_shape(i);
+        new_sliced_size *= Y[i];
+      }
+    }
+    // at least one index needs to be adjustable
+    OP_REQUIRES(context, found, tensorflow::errors::InvalidArgument(
+                                    "periodic_intersperse expects at least "
+                                    "one index to be marked as adjustable."));
+
+    int count = 0;
+    for (const auto dim_info : input_tensor.shape()) {
+      X[count] = dim_info.size;
+      total_size *= X[count];
+      ++count;
+    }
+
+    Y[q] = tensorflow::int64(
+        std::floor(float(total_size) / float(new_sliced_size)));
+
+    count = 0;
+    for (const auto dim_info : input_tensor.shape()) {
+      f[count] = float(Y[count]) / float(X[count]);
+      g[count] = tensorflow::int64(std::ceil(f[count]));
+      if (count == 0)
+        G[count] = 1;
+      else
+        G[count] = G[count - 1] * g[count - 1];
+      ++count;
+    }
+  }
+
+  // ensure that the new dimension is greater than zero
+  OP_REQUIRES(context, Y[q] > 0,
+              tensorflow::errors::InvalidArgument(
+                  "periodic_intersperse found that the "
+                  "adjustable dimension, ",
+                  q, ", isn't greater than zero, ", Y[q], "."));
+  for (int i = 0; i < rank; ++i) {
+    output_shape.AddDim(Y[i]);
+  }
+  const auto new_size = new_sliced_size * Y[q];
+
+  // Create an output tensor and attach it to the current context
+  tensorflow::Tensor* output_tensor = nullptr;
+  OP_REQUIRES_OK(context,
+                 context->allocate_output(0, output_shape, &output_tensor));
+  auto output = output_tensor->flat<T>();
+
+  {
+    // memory is allocated for these variables outside the inner loop for
+    // efficiency (yes, I know I could create a separate class scope for
+    // this purpose instead)
+    typename std::decay<decltype(new_size)>::type result = 0;
```

Comment:
Why not just go with int64 (or size_t) and get rid of all this type inference stuff? It is not very readable, and doesn't really buy you anything here.